common/src/main/java/io/netty/util/concurrent/DefaultPromise.java

```diff
@@ -63,22 +68,51 @@
     private short waiters;
 
     /**
-     * Creates a new instance.
+     * Creates a new unfulfilled promise.
      *
      * It is preferable to use {@link EventExecutor#newPromise()} to create a new promise
      *
      * @param executor
-     *        the {@link EventExecutor} which is used to notify the promise once it is complete.
+     *        The {@link EventExecutor} which is used to notify the promise once it is complete.
      *        It is assumed this executor will protect against {@link StackOverflowError} exceptions.
      *        The executor may be used to avoid {@link StackOverflowError} by executing a {@link Runnable} if the stack
      *        depth exceeds a threshold.
-     *
      */
     public DefaultPromise(EventExecutor executor) {
         this.executor = requireNonNull(executor, "executor");
         stage = new DefaultFutureCompletionStage<>(this);
     }
 
+    /**
+     * Creates a new promise that has already been completed successfully.
+     *
+     * @param executor
+     *        The {@link EventExecutor} which is used to notify the promise once it is complete.
+     *        It is assumed this executor will protect against {@link StackOverflowError} exceptions.
+     *        The executor may be used to avoid {@link StackOverflowError} by executing a {@link Runnable} if the stack
+     *        depth exceeds a threshold.
+     */
+    public DefaultPromise(EventExecutor executor, Object result) {
+        this.executor = requireNonNull(executor, "executor");
+        this.result = result == null? SUCCESS : result;
+        stage = new DefaultFutureCompletionStage<>(this);
+    }
+
+    /**
+     * Creates a new promise that has already failed.
+     *
+     * @param executor
+     *        The {@link EventExecutor} which is used to notify the promise once it is complete.
+     *        It is assumed this executor will protect against {@link StackOverflowError} exceptions.
+     *        The executor may be used to avoid {@link StackOverflowError} by executing a {@link Runnable} if the stack
+     *        depth exceeds a threshold.
+     */
+    public DefaultPromise(Throwable cause, EventExecutor executor) {
```

Comment:
This argument ordering difference seems to just be asking for bugs with code calling wrong constructor. These should probably be static factory methods in DefaultPromise for the failed/completed cases so that the `EventExecutor` is always first and the method name makes it clear what you are constructing.

Reply:
Static methods on classes show up on instances, though. That's why I initially put those static methods on the interfaces.

Reply:
Static method on DefaultPromise seems strongly better than this easy-to-use-wrong-constructor. The chance of calling `existingPromise.newFailedPromise()` seems rather low, especially since many callsites will use Promise-typed variables, and if it happens it causes almost no harm. I believe there's multiple linters that notice issues like that.